Remove project update blocking

fosterseth opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
A running job that has a project update will block that update from running. This fix removes the block.

Adds a functional test that sets up a job in "running" state, and starts a project update that is in "pending" state. Using `TaskManager::is_job_blocked` and `DependenceGraph::is_job_blocked`, assert that the project update is allowed to run.

issue #5153

<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 9.0.1
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

ryanpetrello commented:
@omgjlk you interested in giving this a whirl?

AlanCoding commented:
I'm going to have to get help understanding what this is changing exactly.

But if we do move ahead with this, I think we should lose `JOB_PROJECT_IDS`, because this is the only thing that used it.

omgjlk commented:
I should be able to test this, hopefully today.

matburt commented on awx/main/tests/functional/task_management/test_scheduler.py:
```diff
@@ -214,6 +214,27 @@ def test_single_job_dependencies_project_launch(default_instance_group, job_temp
 
 
+@pytest.mark.django_db
+def test_job_not_blocking_project_update(default_instance_group, job_template_factory):
```
I'm concerned about using `start_task`  being the test point for this. There's a lot of potential for `start_task` itself to bail out (for things like pre-start checks) and it's not unreasonable to think that at some point it might be refactored in such a way that it masks a job dependency problem.

It might be better to hook into the `TaskManager::is_job_blocked` mechanism here, as well as calling out to `DependencyGraph::is_job_blocked` directly to make sure both code paths are covered independently.

fosterseth replied:
@matburt I've updated the test. I now call is_job_blocked from both the TaskManager and DependecyGraph, to check them independently.

matburt reviewed (changes requested):
The logic looks sound, just maybe need to adjust the unit test strategy.

fosterseth commented:
> I'm going to have to get help understanding what this is changing exactly.
> 
> But if we do move ahead with this, I think we should lose `JOB_PROJECT_IDS`, because this is the only thing that used it.

AlanCoding commented on awx/main/scheduler/dependency_graph.py:
```diff
@@ -88,8 +88,7 @@ def mark_workflow_job(self, job):
         self.data[self.WORKFLOW_JOB_TEMPLATES_JOBS][job.workflow_job_template_id] = False
 
     def can_project_update_run(self, job):
-        return self.data[self.JOB_PROJECT_IDS].get(job.project_id, True) and \
```
just to clarify, I think we should remove _all_ references to the variable `JOB_PROJECT_IDS` because it's no longer needed.

fosterseth replied:
removed!

AlanCoding commented:
> A running job that has a project update will block that update from running. This fix removes the block.

I still think we need some docs for this that are a little bit more verbose. How do we write a release note for someone who has no prior knowledge of our task manager blocking behaviors?

#### To describe prior behavior

If a project was manually updated while a job using that project was in progress, the project update would stay in the pending state until the job is finished.

Is that correct?

To draw it out a little bit realistically, say I have a job that runs for 2 hours. Now I make a commit to source control and want to run a _totally different playbook_ from that project with a totally different job template. I have to wait until that 2 hours is up before the server will do the update to pick up my changes?

omgjlk commented:
> 
> To draw it out a little bit realistically, say I have a job that runs for 2 hours. Now I make a commit to source control and want to run a _totally different playbook_ from that project with a totally different job template. I have to wait until that 2 hours is up before the server will do the update to pick up my changes?

This is the reality today. I frequently have long many hour running jobs. While that job is running, I'm 100% blocked from updating the project repo, which would be necessary to start a job with a newly added template, or a newly modified template.

fosterseth committed:
Prevent running jobs from blocking project updates

A running job that has a project update will block
that update from running. This fix removes
the block.
Adds a functional test that sets up a job in "running" state, and
starts a project update that is in "pending" state. Assert that
the task manager and dependency graph .is_job_blocked methods both
return False.

issue #5153

fosterseth commented:
recheck
